modules/frontend/frontend.go

```diff
@@ -65,63 +67,81 @@ func New(cfg Config, next http.RoundTripper, o overrides.Interface, reader tempo
 		return nil, fmt.Errorf("query backend after should be less than or equal to query ingester until")
 	}
 
+	if cfg.Metrics.Sharder.ConcurrentRequests <= 0 {
+		return nil, fmt.Errorf("frontend metrics concurrent requests should be greater than 0")
+	}
+
+	if cfg.Metrics.Sharder.TargetBytesPerRequest <= 0 {
+		return nil, fmt.Errorf("frontend metrics target bytes per request should be greater than 0")
+	}
+
+	if cfg.Metrics.Sharder.Interval <= 0 {
+		return nil, fmt.Errorf("frontend metrics interval should be greater than 0")
+	}
+
 	retryWare := pipeline.NewRetryWare(cfg.MaxRetries, registerer)
 	cacheWare := pipeline.NewCachingWare(cacheProvider, cache.RoleFrontendSearch, logger)
 	statusCodeWare := pipeline.NewStatusCodeAdjustWare()
 	traceIDStatusCodeWare := pipeline.NewStatusCodeAdjustWareWithAllowedCode(http.StatusNotFound)
 
 	tracePipeline := pipeline.Build(
-		[]pipeline.AsyncMiddleware[*http.Response]{
+		[]pipeline.AsyncMiddleware[combiner.PipelineResponse]{
 			multiTenantMiddleware(cfg, logger),
 			newAsyncTraceIDSharder(&cfg.TraceByID, logger),
 		},
 		[]pipeline.Middleware{traceIDStatusCodeWare, retryWare},
 		next)
 
 	searchPipeline := pipeline.Build(
-		[]pipeline.AsyncMiddleware[*http.Response]{
+		[]pipeline.AsyncMiddleware[combiner.PipelineResponse]{
 			multiTenantMiddleware(cfg, logger),
 			newAsyncSearchSharder(reader, o, cfg.Search.Sharder, logger),
 		},
 		[]pipeline.Middleware{cacheWare, statusCodeWare, retryWare},
 		next)
 
 	searchTagsPipeline := pipeline.Build(
-		[]pipeline.AsyncMiddleware[*http.Response]{
+		[]pipeline.AsyncMiddleware[combiner.PipelineResponse]{
 			multiTenantMiddleware(cfg, logger),
 			newAsyncTagSharder(reader, o, cfg.Search.Sharder, parseTagsRequest, logger),
 		},
 		[]pipeline.Middleware{cacheWare, statusCodeWare, retryWare},
 		next)
 
 	searchTagValuesPipeline := pipeline.Build(
-		[]pipeline.AsyncMiddleware[*http.Response]{
+		[]pipeline.AsyncMiddleware[combiner.PipelineResponse]{
 			multiTenantMiddleware(cfg, logger),
 			newAsyncTagSharder(reader, o, cfg.Search.Sharder, parseTagValuesRequest, logger),
 		},
 		[]pipeline.Middleware{cacheWare, statusCodeWare, retryWare},
 		next)
 
+	// metrics summary
 	metricsPipeline := pipeline.Build(
-		[]pipeline.AsyncMiddleware[*http.Response]{
+		[]pipeline.AsyncMiddleware[combiner.PipelineResponse]{
 			multiTenantUnsupportedMiddleware(cfg, logger),
 		},
 		[]pipeline.Middleware{statusCodeWare, retryWare},
 		next)
 
-	queryRangeMiddleware := MergeMiddlewares(
-		newMultiTenantUnsupportedMiddleware(cfg, logger),
-		newQueryRangeMiddleware(cfg, o, reader, logger), retryWare)
+	// traceql metrics
+	queryRangePipeline := pipeline.Build(
+		[]pipeline.AsyncMiddleware[combiner.PipelineResponse]{
+			multiTenantMiddleware(cfg, logger),
```

Comment:
Is this enabling multi-tenant support?  I think it would work in the current setup (combiner), but wondering if we should hold on this until we support the next round of features.

Reply:
Yup, this is enabling multitenant. I'm fine with going back to unsupported. Does this impact the metrics sharding?

Reply:
Sharding is ok, I had in mind the changes needed to the sharder/combiner for histograms. But after more consideration it is good.